### ISSUE TITLE
Spinning in 3D preview is happening again

### ISSUE THREAD

**Issue by JPLeBreton** [sirjuddington/SLADE#1595]
**SLADE Version**: 3.2.4
**Operating System**: Linux (Manjaro, stable branch)

**Issue Details**:
Steps to reproduce:
1. Open a map in Slade's map editor
2. Press Q to enter the 3D preview mode
3. Observe: the view is looking straight down at the ground and spinning (rolling). Mouse and keyboard input can change the orientation but the spinning can't be stopped, making it impossible to work in the 3D view.

I know this has been reported and [fixed](https://github.com/sirjuddington/SLADE/issues/1048) a couple times already in the history of the project, but it's started happening again. I'm on a Linux system with an Nvidia GPU (535 drivers) on a GNOME desktop with the Wayland display server. Every other aspect of Slade runs flawlessly, including the 2D map editor view.

**Comment by eGax** [comment]
That issue was referencing MAC builds with spinning issues though.

**Comment by JPLeBreton** [comment]
> Looks like this is another one of those Wayland issues - apparently mouse pointer warping (which 3d mode mouselook relies on to work) is not implemented at all in Wayland (see [wxWidgets/wxWidgets#23778](https://github.com/wxWidgets/wxWidgets/issues/23778)). Not sure if I can do anything about it until it's added unfortunately, as I don't think there is any other way to get mouselook to work using wxWidgets.

Thanks for looking into this, I'll see if there's anything I can do to coax the support along upstream.

**Comment by iamtoaster** [comment]
Maybe it is possible to implement this using GTK directly with some Wayland protocol interactions? I made a proof of concept in #1801 that implements exactly that. It adds an optional dependency on wayland libraries though, and make dependency on GTK3 explicit. I am unsure if it will even work with GTK2. I also did not really think much about how it should fit into existing codebase, so this will definitely need a lot of love. But it works. Kinda. My POC makes the camera movements in 3D view jerky, and I am unsure if it is supposed to be like that. @sirjuddington

**Comment by Blzut3** [comment]
Camera movements are smooth for me under XWayland.  (Probably goes without saying but this is without your change set applied.)

**Comment by Person-Who-Exists** [comment]
There is a new Wayland protocol in the works that would enable cursor warping which should fix this issue.

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/337

**Comment by JPLeBreton** [comment]
Looks like the relevant pointer warping support has been merged into wx: https://github.com/wxWidgets/wxWidgets/issues/23778

**Comment by joanbm** [comment]
It seems unlikely that the wxWidgets 3.3.2 release will solve this in any effective capacity - [according to this page, the wp_pointer_warp_v1 Wayland protocol isn't supported by any mainstream Wayland compositor](https://wayland.app/protocols/wayland-protocols/337#compositor-support).

Based on (https://github.com/wxWidgets/wxWidgets/issues/23778#issuecomment-1719523909) it could also be implemented via [zwp_pointer_constraints_v1](https://wayland.app/protocols/pointer-constraints-unstable-v1#zwp_pointer_constraints_v1) but wxWidgets doesn't currently use it or expose any API that does.

CC @Blzut3 due to https://github.com/sirjuddington/SLADE/issues/1351#issuecomment-3420583837.

**Comment by Blzut3** [comment]
KDE is supposed to support in the imminent 6.5 release: https://blogs.kde.org/2025/07/26/this-week-in-plasma-printer-ink-level-monitoring/#plasma-650-3

**Comment by joanbm** [comment]
Eureka! I was able to get the 3D map editor camera to work on KDE 6.5 (Wayland), wxWidgets from git, and SLADE with the following change to warp the pointer using wxWidgets instead of SFML:

```diff
diff --git a/src/MapEditor/UI/MapCanvas.cpp b/src/MapEditor/UI/MapCanvas.cpp
index d1ca8695..667539f4 100644
--- a/src/MapEditor/UI/MapCanvas.cpp
+++ b/src/MapEditor/UI/MapCanvas.cpp
@@ -116,9 +116,9 @@ void MapCanvas::draw()
 // -----------------------------------------------------------------------------
 void MapCanvas::mouseToCenter()
 {
-	auto rect   = GetScreenRect();
+	auto size = GetClientSize();
 	mouse_warp_ = true;
-	sf::Mouse::setPosition(sf::Vector2i(rect.x + int(rect.width * 0.5), rect.y + int(rect.height * 0.5)));
+	WarpPointer(size.GetWidth() / 2, size.GetHeight() / 2);
 }
 
 // -----------------------------------------------------------------------------
```

(The patch is likely required, since I believe screen absolute warping won't be supported in Wayland)

I'll test some more before submitting a PR since I'm afraid of it causing regressions on other platforms.

**Comment by joanbm** [comment]
It should also work on GNOME since support was added in Mutter 49 ([1](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4525), [2](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4626), [3](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4678)) but it ends up failing [this check](https://gitlab.gnome.org/GNOME/mutter/-/blob/fae3d35366831e089e42570ca73ad60d62b08d4b/src/wayland/meta-wayland-pointer.c#L1477-1486), possibly a wxWidgets issue.

**Comment by joanbm** [comment]
Since [this commit](https://gitlab.gnome.org/GNOME/mutter/-/commit/83b4b078a94a5dcaad0da618b3a310638b2efa00) in Mutter, it's works in GNOME with some jank:
- Even when not moving the mouse, the camera sometimes moves slightly.
- If using "Enter 3D mode at mouse cursor" from the right click menu, it still spins until you click. This is because Mutter requires for the surface to be clicked before allowing warping the pointer.

It's at least minimally usable, though.

**Comment by Blzut3** [comment]
Probably the best we'll get to hope for from GNOME, although to be fair to them Slade's use case is the one that pointer constraints exists for.  The separation in purpose doesn't really mesh with the cross platform frameworks.

**Comment by joanbm** [comment]
Yeah, I'm glad there's at least a path for it to minimally work, but I'm wary that it may never be as good as it was in X11.

From the SLADE side we could likely work around by making the camera move via "click and drag" instead of pointer capture, but in my view ideally a solution should come from upper layers (wxWidgets + Wayland) rather than SLADE.

**Comment by sirjuddington** [comment]
"Click and drag" camera movement is exactly what I'm experimenting with in the 3.3.0 3d mode implementation - it'd come with the advantage of being able to select walls/flats using the mouse pointer instead of having to point the camera at them directly, and of course working around mouse capture issues. It'll still likely have the existing camera control as an option, or perhaps a toggle by pressing a key.
